### PR TITLE
Fix crash on rest element destructuring with null initializer

### LINKED ISSUE #2087: Compiler crash on rest destructuring param with null initializer
```ts
function foo([...r] = null){}
```

TypeError: Cannot read property 'id' of undefined
    at getTypeListId (D:\GitHub\TypeScript\built\local\tsc.js:9833:41)
    at createTypeReference (D:\GitHub\TypeScript\built\local\tsc.js:9854:22)
    at createArrayType (D:\GitHub\TypeScript\built\local\tsc.js:9973:52)
    at getTypeForBindingElement (D:\GitHub\TypeScript\built\local\tsc.js:8886:32)
    at getTypeForVariableLikeDeclaration (D:\GitHub\TypeScript\built\local\tsc.js:8896:24)
    at getWidenedTypeForVariableLikeDeclaration (D:\GitHub\TypeScript\built\local\tsc.js:8957:24)
    at getTypeOfVariableOrParameterOrProperty (D:\GitHub\TypeScript\built\local\tsc.js:8987:28)
    at checkVariableLikeDeclaration (D:\GitHub\TypeScript\built\local\tsc.js:14014:24)
    at checkBindingElement (D:\GitHub\TypeScript\built\local\tsc.js:14044:20)
    at checkSourceElement (D:\GitHub\TypeScript\built\local\tsc.js:14909:28)

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #2087.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1722,7 +1722,7 @@ module ts {
             }
             else {
                 // For an array binding element the specified or inferred type of the parent must be assignable to any[]
-                if (!isTypeAssignableTo(parentType, anyArrayType)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Is `anyArrayType` used anywhere else?

**Comment by DanielRosenwasser** [comment]
Consider adding a test using `undefined` as the initializer as well.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1722,7 +1722,7 @@ module ts {
             }
             else {
                 // For an array binding element the specified or inferred type of the parent must be assignable to any[]
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: This comment no longer reflects the code below it.

**Comment by CyrusNajmabadi** [comment]
:+1:   (with the comment fix)

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Yes, it is used in the new `isArrayLikeType` function, plus some other places.

**Pull request merged**